### DISCUSSION TITLE
Nesting shortcodes or triple colon

### DISCUSSION THREAD

**Discussion by OskarKlintrot** [statiqdev/Discussions#212] [Q&A] [answered]
My end goal is to get this html, roughly:

```html
<div class="recipe">
  <div class="ingredients">...</div>
  <div class="instructions">...</div>
</div>
```

Triple colon gives me div's with classes, so that felt like a good place to start:

```markdown
::: recipe

::: ingredients

- list
- of
- stuff

:::

::: instructions

1. how
1. to

:::

:::
```

That almost worked but the second div ended up as a siblings to the first instead of a child:

![image](https://github.com/statiqdev/Discussions/assets/6783879/f5d87826-bd1a-4e87-ad1d-65c641ea1a01)

Let's make the first triple colon a simple shortcode instead:

```csharp
public sealed class RecipeShortcode : SyncShortcode
{
    private const string Class = nameof(Class);

    public override ShortcodeResult Execute(KeyValuePair<string, string>[] args, string content, IDocument document, IExecutionContext context)
    {
        XElement div = new(
            "div",
            content,
            new XAttribute(Class, "recipe"));

        return div.ToString();
    }
}
```

```markdown
<?! recipe ?>

::: ingredients

- list
- of
- stuff

:::

::: instructions

1. how
1. to

:::

<?!/ recipe ?>
```

That works like a charm!

![image](https://github.com/statiqdev/Discussions/assets/6783879/d4d25d77-2e9c-44e8-ac2d-fe1d8b712442)

However, now I want to use the `Include` shortcode and then it seems as if my `Recipe` shortcode escapes the inner shortcode:

![image](https://github.com/statiqdev/Discussions/assets/6783879/dc30664d-ff3d-4a22-badd-294a7b8d5bdf)

How can I solve this? I tried to look through the built in shortcodes for hints but I'm still stuck at this.

______________
Update: I found [this answer on SO](https://stackoverflow.com/a/17762116/4826084) and it works great!

**Comment by OskarKlintrot** [comment, marked as answer]
```csharp

using System.Xml.Linq;

namespace My.App;

public sealed class RecipeShortcode : SyncShortcode
{
    private const string Class = nameof(Class);

    public override ShortcodeResult Execute(KeyValuePair<string, string>[] args, string content, IDocument document, IExecutionContext context)
    {
        XElement div = new(
            "div",
            new XRaw(content),
            new XAttribute(Class, "recipe"));

        return div.ToString();
    }

    private sealed class XRaw(string text) : XText(text)
    {
        public override void WriteTo(System.Xml.XmlWriter writer) => writer.WriteRaw(Value);
    }
}

```

**Reply by daveaglick** [reply]
That's awesome! I haven't documented or promoted shortcodes nearly enough, so I'm thrilled you got a custom one to work for you. Hopefully 2024 is the year I get all these "hidden" features out there (and probably fix a bunch of bugs as folks start using them more heavily).
